core/slim_sim.cpp

```diff
@@ -6191,13 +6202,111 @@ void SLiMSim::WritePopulationTable(tsk_table_collection_t *p_tables)
 	}
 }
 
+void SLiMSim::WriteTreeSequenceMetadata(tsk_table_collection_t *p_tables)
+{
+    int ret = 0;
+
+    //////
+    // Top-level (tree sequence) metadata:
+    // In the future, we might need to *add* to the metadata *and also* the schema,
+    // leaving other keys that might already be there.
+    // But that's being a headache, so we're skipping it.
+    nlohmann::json metadata;
+	metadata["SLiM"]["model_type"] = (ModelType() == SLiMModelType::kModelTypeWF) ? "WF" : "nonWF";
+	metadata["SLiM"]["generation"] = Generation();
+	metadata["SLiM"]["file_version"] = SLIM_TREES_FILE_VERSION;
+    if (spatial_dimensionality_ == 0) {
+        metadata["SLiM"]["spatial_dimensionality"] = "";
```

Comment:
Oh, hm. Well, right now I don't have *any* of the keys listed as "required". But I think we should just make them all required - it's easier to just check the value rather than first checking if the key is present, then checking its value... what do you think?

Reply:
Hmm.  One the one hand, requiring a key that no software actually requires seems iffy; it elides the distinction between "required" and "optional" just for a bit of convenience.  On the other hand, it *is* convenient to be guaranteed that a key must be present, and be able to assume it, yes.  Is that how it would actually play out on the Python side?  I.e., suppose someone writes a file that is missing a key that is marked as required.  When that file gets read in on the Python side, does the missing key automatically get flagged as a fatal error before the client code that requested the ts to be loaded ever gets back control?  Or would the client code really still need to check that the "required" key was actually present, in order to avoid some kind of cryptic runtime error?  If the answer is yes, the Python tskit code enforces required keys automatically, then I'm on the fence about this, do as you think best.  If it doesn't enforce that, then there's no point in marking the key as "required" for convenience, since in practice you'd still have to check that it's actually there.

Reply:
I think that although we don't use these things currently, I will very soon want to write methods in pyslim that do use these things. So, let's require them. And, yes - if you try to load a tree sequence with malformed metadata (ie that doesn't match the schema) then tskit throws an error.

Reply:
OK, works for me.